## line imports NumPy and Pandas, essential libraries for data analysis and handling missing values.


In [226]:
import numpy as np
import pandas as pd
import seaborn as sns

##  Load the Netflix dataset into a pandas DataFrame

In [227]:

df = pd.read_csv(r"C:\Users\mouni\OneDrive\Desktop\netflix_titles.csv\netflix_titles.csv")
df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

## Displays summary info about the DataFrame: column names, non-null counts, and data types there are 8806 rows are there


In [228]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


## Returns a DataFrame of the same shape with True for missing values and False otherwise

In [229]:
df.isnull()

show_id   type  title  director   cast  country  date_added  \
0       False  False  False     False   True    False       False   
1       False  False  False      True  False    False       False   
2       False  False  False     False  False     True       False   
3       False  False  False      True   True     True       False   
4       False  False  False      True  False    False       False   
...       ...    ...    ...       ...    ...      ...         ...   
8802    False  False  False     False  False    False       False   
8803    False  False  False      True   True     True       False   
8804    False  False  False     False  False    False       False   
8805    False  False  False     False  False    False       False   
8806    False  False  False     False  False    False       False   

      release_year  rating  duration  listed_in  description  
0            False   False     False      False        False  
1            False   False     False      False        False  
2            False   False     False      False        False  
3            False   False     False      False        False  
4            False   False     False      False        False  
...            ...     ...       ...        ...          ...  
8802         False   False     False      False        False  
8803         False   False     False      False        False  
8804         False   False     False      False        False  
8805         False   False     False      False        False  
8806         False   False     False      False        False  

[8807 rows x 12 columns]

## The .mode() function in Pandas returns the most frequent value(s) in a Series (column). It's often used to fill missing values with the most common entry

In [230]:
df.mode()

show_id   type                          title       director  \
0         s1  Movie                         #Alive  Rajiv Chilaka   
1        s10    NaN  #AnneFrank - Parallel Stories            NaN   
2       s100    NaN              #FriendButMarried            NaN   
3      s1000    NaN            #FriendButMarried 2            NaN   
4      s1001    NaN                          #Roxy            NaN   
...      ...    ...                            ...            ...   
8802    s995    NaN                       忍者ハットリくん            NaN   
8803    s996    NaN                           海的儿子            NaN   
8804    s997    NaN                           마녀사냥            NaN   
8805    s998    NaN                        반드시 잡는다            NaN   
8806    s999    NaN            최강전사 미니특공대 : 영웅의 탄생            NaN   

                    cast        country       date_added  release_year rating  \
0     David Attenborough  United States  January 1, 2020        2018.0  TV-MA   
1                    NaN            NaN              NaN           NaN    NaN   
2                    NaN            NaN              NaN           NaN    NaN   
3                    NaN            NaN              NaN           NaN    NaN   
4                    NaN            NaN              NaN           NaN    NaN   
...                  ...            ...              ...           ...    ...   
8802                 NaN            NaN              NaN           NaN    NaN   
8803                 NaN            NaN              NaN           NaN    NaN   
8804                 NaN            NaN              NaN           NaN    NaN   
8805                 NaN            NaN              NaN           NaN    NaN   
8806                 NaN            NaN              NaN           NaN    NaN   

      duration                     listed_in  \
0     1 Season  Dramas, International Movies   
1          NaN                           NaN   
2          NaN                           NaN   
3          NaN                           NaN   
4          NaN                           NaN   
...        ...                           ...   
8802       NaN                           NaN   
8803       NaN                           NaN   
8804       NaN                           NaN   
8805       NaN                           NaN   
8806       NaN                           NaN   

                                            description  
0     Paranormal activity at a lush, abandoned prope...  
1                                                   NaN  
2                                                   NaN  
3                                                   NaN  
4                                                   NaN  
...                                                 ...  
8802                                                NaN  
8803                                                NaN  
8804                                                NaN  
8805                                                NaN  
8806                                                NaN  

[8807 rows x 12 columns]

## Handling duplicate values


In [231]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8802    False
8803    False
8804    False
8805    False
8806    False
Length: 8807, dtype: bool

## This line fills missing values in the ‘type’ column using the most frequently occurring country (mode)

In [232]:
df["type"].mode()

0    Movie
Name: type, dtype: object

In [233]:
df['type'] = df['type'].fillna(df['type'].mode())
df['type']

0         Movie
1       TV Show
2       TV Show
3       TV Show
4       TV Show
         ...   
8802      Movie
8803    TV Show
8804      Movie
8805      Movie
8806      Movie
Name: type, Length: 8807, dtype: object

## This line fills missing values in the ‘director’ column using the most frequently occurring country (mode)

In [234]:
df['director'].mode()

0    Rajiv Chilaka
Name: director, dtype: object

## Replace the value into mode value

In [235]:
df['director'] = df['director'].fillna(df['director'].mode())
df['director']

0       Kirsten Johnson
1                   NaN
2       Julien Leclercq
3                   NaN
4                   NaN
             ...       
8802      David Fincher
8803                NaN
8804    Ruben Fleischer
8805       Peter Hewitt
8806        Mozez Singh
Name: director, Length: 8807, dtype: object

## This line fills missing values in the ‘cast’ column using the most frequently occurring country (mode)

In [236]:
df['cast'].mode()

0    David Attenborough
Name: cast, dtype: object

In [237]:
df['cast'].fillna('David Attenborough',inplace=True)

## This line fills missing values in the ‘country’ column using the most frequently occurring country (mode)

In [238]:
df['country'].mode()

0    United States
Name: country, dtype: object

In [239]:
df['country'] = df['country'].fillna(df['country'].mode())
df['country']

0       United States
1        South Africa
2                 NaN
3                 NaN
4               India
            ...      
8802    United States
8803              NaN
8804    United States
8805    United States
8806            India
Name: country, Length: 8807, dtype: object

In [240]:
df.isnull()

show_id   type  title  director   cast  country  date_added  \
0       False  False  False     False  False    False       False   
1       False  False  False      True  False    False       False   
2       False  False  False     False  False     True       False   
3       False  False  False      True  False     True       False   
4       False  False  False      True  False    False       False   
...       ...    ...    ...       ...    ...      ...         ...   
8802    False  False  False     False  False    False       False   
8803    False  False  False      True  False     True       False   
8804    False  False  False     False  False    False       False   
8805    False  False  False     False  False    False       False   
8806    False  False  False     False  False    False       False   

      release_year  rating  duration  listed_in  description  
0            False   False     False      False        False  
1            False   False     False      False        False  
2            False   False     False      False        False  
3            False   False     False      False        False  
4            False   False     False      False        False  
...            ...     ...       ...        ...          ...  
8802         False   False     False      False        False  
8803         False   False     False      False        False  
8804         False   False     False      False        False  
8805         False   False     False      False        False  
8806         False   False     False      False        False  

[8807 rows x 12 columns]

## Checking any duplicate values are required or not

In [241]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8802    False
8803    False
8804    False
8805    False
8806    False
Length: 8807, dtype: bool

## converting the format string to int

In [242]:
df['date_added'] = pd.to_datetime(df['date_added'], format='mixed', dayfirst=False, errors='coerce')
df['date_added']

0      2021-09-25
1      2021-09-24
2      2021-09-24
3      2021-09-24
4      2021-09-24
          ...    
8802   2019-11-20
8803   2019-07-01
8804   2019-11-01
8805   2020-01-11
8806   2019-03-02
Name: date_added, Length: 8807, dtype: datetime64[ns]

In [243]:
df['date_added']=pd.to_datetime(df['date_added'])
df['date_added']

0      2021-09-25
1      2021-09-24
2      2021-09-24
3      2021-09-24
4      2021-09-24
          ...    
8802   2019-11-20
8803   2019-07-01
8804   2019-11-01
8805   2020-01-11
8806   2019-03-02
Name: date_added, Length: 8807, dtype: datetime64[ns]

## Converting the format string to int

In [244]:
df['date_added'] = pd.to_datetime(df['date_added'], format='mixed', dayfirst=False, errors='coerce')
df['date_added']

0      2021-09-25
1      2021-09-24
2      2021-09-24
3      2021-09-24
4      2021-09-24
          ...    
8802   2019-11-20
8803   2019-07-01
8804   2019-11-01
8805   2020-01-11
8806   2019-03-02
Name: date_added, Length: 8807, dtype: datetime64[ns]

## to conert the format into date/month/year 

In [245]:
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')
df['date_added'] = df['date_added'].dt.strftime('%d-%m-%Y')
df['date_added']


0       25-09-2021
1       24-09-2021
2       24-09-2021
3       24-09-2021
4       24-09-2021
           ...    
8802    20-11-2019
8803    01-07-2019
8804    01-11-2019
8805    11-01-2020
8806    02-03-2019
Name: date_added, Length: 8807, dtype: object

## Capitalize the first letter of each column name

In [246]:
df.columns = df.columns.str.capitalize()  


In [247]:
df

Show_id     Type                  Title         Director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   Cast        Country  \
0                                    David Attenborough  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                    David Attenborough            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                 David Attenborough            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

      Date_added  Release_year Rating   Duration  \
0     25-09-2021          2020  PG-13     90 min   
1     24-09-2021          2021  TV-MA  2 Seasons   
2     24-09-2021          2021  TV-MA   1 Season   
3     24-09-2021          2021  TV-MA   1 Season   
4     24-09-2021          2021  TV-MA  2 Seasons   
...          ...           ...    ...        ...   
8802  20-11-2019          2007      R    158 min   
8803  01-07-2019          2018  TV-Y7  2 Seasons   
8804  01-11-2019          2009      R     88 min   
8805  11-01-2020          2006     PG     88 min   
8806  02-03-2019          2015  TV-14    111 min   

                                              Listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            Description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]